drivers/mtd/mtd_partition.c

```diff
@@ -413,6 +413,25 @@ static int part_ioctl(FAR struct mtd_dev_s *dev, int cmd, unsigned long arg)
         }
         break;
 
+      case MTDIOC_PARTINFO:
+        {
+          FAR struct partition_info_s *info =
+            (FAR struct partition_info_s *)arg;
+          if (info != NULL)
+            {
+              info->magic       = 0;
+              info->numsectors  = priv->neraseblocks * priv->blkpererase;
+              info->length      = info->numsectors * priv->blocksize;
+              info->sectorsize  = priv->blocksize;
+              info->startsector = priv->firstblock;
+              info->endsector   = priv->firstblock + info->numsectors;
+              info->parent      = priv->parent->name;
+
+              ret = OK;
+          }
+        }
+        break;
+
```

Comment:
This is redundant.  This information is all available in the procfs/.  See https://github.com/apache/incubator-nuttx/blob/master/drivers/mtd/mtd_partition.c#L617

Reply:
Yes, but how should the application consume this information from `procfs`? Please, correct if I misunderstood, but a read access via `procfs` retrieves the information as string. So the application should parse this info. If so, this is sub-optimal for an application that just requires the a single information, don't you think?

Reply:
> Please, correct if I misunderstood it, but a read access via `procfs` retrieves the information as string. So the application should parse this info. If so, this is sub-optimal for an application that just requires a single information, don't you think?

We had that discussion years ago.  When Ken created procfs, some people want it to be human readable strings (like Linux) and others wanted a binary interface.  It was resolved then that we would use the string interface and that we would pay that price.  This is a resolved issue and there is really no point in raising it again.

Reply:
It seems like a VERY odd technical decision to duplicate an existing interface purely to satisfy your personal preference.  Redundancy like that should not be tolerated in an OS and has never been tolerated in the past.

Reply:
One thing that I am not sure that I understand is how to associate a partition number with a block or MTD driver.  In Linux, that might be done with the minor number of the device like /dev/sdaN where N is the partition number.  But there are no real minor numbers in NuttX although such a convention could be adopted.

Reply:
> It seems like a VERY odd technical decision to duplicate an existing interface purely to satisfy your personal preference. Redundancy like that should not be tolerated in an OS and has never been tolerated in the past.

It's sad to hear that you believe I am bringing personal motivations to this PR. My original intent was to just provide one single piece of data to the application (the `firstblock` info pro MTD partition), and I am dedicating my time to work with you after each round of discussions, with the sole purpose of trying to do what's best to the project.
So, please, I'd ask you to be a bit more reasonable.

Reply:
> Redundancy like that should not be tolerated in an OS and has never been tolerated in the past.

There has always been at least two methods for retrieving information from MTD:
- `ioctl` commands from the `MTDIOC_*`;
- `procfs`;

Currently neither approach provides the `firstblock` information.

So, consider an application that already owns the path to the MTD character device (via BCH and FTL) and already manipulates the MTD partition via file descriptors. So this is not one particular use case, it is a fairly common one.
Linux also provides this information via `ioctl` calls.

`procfs` requires `CONFIG_FS_PROCFS`, which is an optional feature. Adding `CONFIG_FS_PROCFS`, even after disabling all `procfs` entries except for the **partitions**, this imposes a footprint overhead of 2460 bytes to the binary file. There is also the runtime execution overhead of the two-way string conversion.

Do you really think it is worth paying the price for this coupling to `procfs`?

Reply:
Some more evidences to justify the `ioctl` path:

1) Linux MTD character device driver implements similar `ioctl` commands (**MEMGETREGIONINFO and **MEMGETINFO**) for providing MTD info to the application:
https://github.com/torvalds/linux/blob/master/drivers/mtd/mtdchar.c#L684-L718

2) `mtd-utils` rely on **MEMGETINFO** for retrieving MTD partition information:
https://github.com/sigma-star/mtd-utils/blob/master/misc-utils/ftl_check.c#L80-L84

Reply:
Following the Linux path is the right way to go!

Reply:
It's good to follow Linux design, two question:

1. Should we also use MEMGETINFO and mtd_info_t too?
2. How about block driver which should have the similar requirement too?

Reply:
> It's good to follow Linux design, two question:
> 
> 1. Should we also use MEMGETINFO and mtd_info_t too?

For compatibility purposes, I think we should.
On Linux `mtd_info_t` is a typedef to `struct mtd_info_user` and some of its fields overlap with NuttX `mtd_geometry_s`. So this would require to refactor `MTDIOC_GEOMETRY` to avoid having two `ioctl` commands returning the same information.
Since this is not a simple task, I'd suggest to refactor it in another PR.

> 2. How about block driver which should have the similar requirement too?

I also agree on this one.
The intent with the proposed `struct partition_info_s` was to make it generic, so that it could be used by block devices also.
But, again, I'd like to restrict the scope of this PR to MTD only to divide the validation effort.

I believe that the proposed `MTDIOC_PARTINFO` does not overlap with `MEMGETINFO` nor `MEMGETREGIONINFO`. As explained by @patacongo, the MTD partition abstraction is more of a software abstraction than a real partition, and it is different to what Linux refers to a MTD Region, which means areas of an MTD with distinct erase sizes, something that we don't have in NuttX so far (to my knowledge).

So merging this proposal would not be a wasted effort. We may later have both `MEMGETINFO` for compatibility purposes by refactoring the MTDIOC_GEOMETRY, and also `MEMGETREGIONINFO` if eventually we support these types of MTD with more than one erase region.

Reply:
>     1. Linux MTD character device driver implements similar `ioctl` commands (**MEMGETREGIONINFO** and **MEMGETINFO**) for providing MTD info to the application:

That argument has value if the solution is 100% compatible with Linux.

Reply:
> That argument has value if the solution is 100% compatible with Linux.

Linux only provides the partition offset information via `sysfs`, a subsystem that NuttX does not currently have.
So 100% compatibility with Linux is not possible at the moment.
But, the approach via `ioctl` is the most compatible to Linux.

Reply:
> So 100% compatibility with Linux is not possible at the moment.
> But, the approach via `ioctl` is the most compatible to Linux.

That makes no sense to me.  That is like saying I need a whale, but I don't have a whale.  But a whale is mammal so I can use a bat instead.

Reply:
Please, explain which approach does make sense to you then.
I really hope that you have a better proposal than the string parsing approach via `procfs`.

Reply:
> I really hope that you have a better proposal than the string parsing approach via `procfs`.

I would prefer in this order:  A 100% Linux compatible solution, string parsing via procfs.  I personally will not merge a non-standard, redundant solution.  But others may.

Reply:
> I would prefer in this order: A 100% Linux compatible solution then string parsing via the existing procfs.

A 100% Linux compatible solution implies the creation of a `sysfs` subsystem for Linux. Although much desirable, unfortunately is far beyond the scope for this PR.
And string parsing via `procfs` when compared to `ioctl` solution is sub-optimal in terms of execution time and code size. Execution time will scale depending on the number of partitions to be returned from `/proc/partitions`. Code size will increase due to the requirement of enabling the optional `procfs` subsystem.
Besides the fact that the only available method for identifying a partition from `/proc/partitions` is via the partition name, which is once again sub-optimal considering it involves comparison of strings.

Using `ioctl` only requires a valid file descriptor and does not impose any architectural changes to the OS, since the MTD subsystem already implements several other `ioctl` commands. Also, the value to be returned via the `ioctl` command is readily available in the MTD partition information.

@xiaoxiang781216 Please kindly evaluate the answers from https://github.com/apache/incubator-nuttx/pull/4248#discussion_r682578813 and also the above reasoning.

Reply:
@gustavonihei it's reasonable to split the big change into the small patch, your propose is fine to me. Thanks for your effort to provide a general solution.